### PR TITLE
Add initial Query construct & SQL generation

### PR DESCRIPTION
**Pull request by wchamber**

### CODE REVIEW

File: api/src/main/java/bio/terra/tanagra/service/search/Query.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+public abstract class Query {
+  /** What to select for the query. */
+  // TODO consider selections with a different entity than the primary entity.
+  public abstract ImmutableList<Selection> selections();
+
+  /** The primary entity being queried. */
+  public abstract Entity primaryEntity();
+
+  /** The filter to apply to the primary entity, if there is a filter. */
+  public abstract Optional<Filter> filter();
```
Comment (calbach) [inline api/src/main/java/bio/terra/tanagra/service/search/Query.java:20]: Looks like this is probably covered by the above TODO, but most of what the dataset builder (specifically via concept sets) is doing currently is applying filters to related entities. But first joined onto an initial cohort definition, i.e. set of person rows defined by a filter.

Not sure how this is modeled here.

File: api/src/main/java/bio/terra/tanagra/service/search/Selection.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.google.auto.value.AutoValue;
+import java.util.Optional;
+
+/**
+ * A construct in a query that represents a value to be selected.
+ *
+ * <p>Represents a single column in a SELECT SQL clause.
+ */
+public interface Selection {
```
Comment (calbach) [inline api/src/main/java/bio/terra/tanagra/service/search/Selection.java:11]: Is there a clear path for "group by" in this model?

**Review by calbach — APPROVED** [review]
Looks good, just a few questions.

File: api/src/main/java/bio/terra/tanagra/service/search/Query.java (hunk 1 above)
Comment (wchamber) [reply api/src/main/java/bio/terra/tanagra/service/search/Query.java:20]: Yep, very much TODO'd. right now I'm expecting it to be a subtype of filter.

File: api/src/main/java/bio/terra/tanagra/service/search/Selection.java (hunk 2 above)
Comment (wchamber) [reply api/src/main/java/bio/terra/tanagra/service/search/Selection.java:11]: Not yet. In many cases, I think we only care about any grouping by at the top most query level, but it's feasible to consider using group by's within a filter.

**Pull request merged**